### PR TITLE
PETSc interface

### PR DESCRIPTION
**Pull request by dajuno**
Parallel implementation of the PETSc interface.

TODO: eigenvalue solver

### CODE REVIEW

File: tests/test_PETSc.py [hunk 1]
```diff
@@ -136,18 +145,22 @@ def test_PETSc(nx=4):
     dof = 4
     ny = nx
     nz = nx
-    parameters = {"Reynolds Number": 0}
+    parameters = {"Reynolds Number": 0, "Verbose": True}
 
-    interface = PETScInterface.Interface(parameters, nx, ny, nz, dim, dof)
+    interface = PETScInterface.Interface(parameters, nx, ny, nz, dim, dof, PETSc.COMM_WORLD)
```
Comment (Sbte) [inline tests/test_PETSc.py:305]: No need to specify this. This is the default.

File: fvm/interface/PETSc.py [hunk 2]
```diff
@@ -34,39 +40,80 @@ class Interface(ParallelBaseInterface):
...
 
-        self.jac.setValuesCSR(
-            numpy.array(local_jac.begA, dtype=numpy.int32),
-            numpy.array(local_jac.jcoA[: local_jac.begA[-1]], dtype=numpy.int32),
-            local_jac.coA[: local_jac.begA[-1]],
-        )
-
-        self.jac.assemble()
+        self.jac.assemblyBegin()
+        if self.comm.size > 1:
```
Comment (Sbte) [inline fvm/interface/PETSc.py]: I think I prefer not to have special cases like this. Cases like this are causing issues in other codes we use where it does work in parallel but not on a single core.

File: tests/test_PETSc.py [hunk 3]
```diff
@@ -113,19 +115,26 @@ def test_norm():
...
+    parameters = {}
 
     state = numpy.zeros(n)
     for i in range(n):
         state[i] = i + 1
 
-    state_dist = PETScInterface.Vector.from_array(state)
+    interface = PETScInterface.Interface(parameters, nx, ny, nz, dim, dof)
+
+    state_dist = PETScInterface.Vector.from_array(interface.map, state)
```
Comment (Sbte) [inline tests/test_PETSc.py:284]: In the Epetra interface I have a method that does it so you don't need to know about the map as a user: https://github.com/BIMAU/transiflow/blob/master/transiflow/interface/Epetra.py#L88

**Review by Sbte — CHANGES_REQUESTED** [review]

**Comment by Sbte** [comment]
I created #28 which fixes the HYMLS tests on multiple cores, in case you're interested what was still failing on that side. After that PR nothing fails anymore.

File: fvm/interface/PETSc.py (hunk 2 above)
Comment (dajuno) [reply fvm/interface/PETSc.py]: Hmm, I think this was a debugging residue. The multi core case also works on a single core. Will delete

**Comment by Sbte** [comment]
I also created #29 to show how I want to integrate the PETSc tests into the continuation tests. That doesn't have to be done in this PR though.

**Comment by Sbte** [comment]
It looks like things go very wrong when running the tests in parallel (everything becoming NaN). That should be fixed.

**Comment by Sbte** [comment]
I debugged it a bit and it seems like `ghostUpdate` is doing the wrong thing.

**Comment by Sbte** [comment]
I think the LGMap is ignored, and that's what is causing the issue.

**Comment by Sbte** [comment]
What I tried now was to run the first ldc test on 4 cores. Then I printed the state vector and the indices (real and ghost) and compared. The state just has increasing numbers in it, so the indices should match the values (-1). They don't, which is what is causing issues. I'm quite sure that the reason is that the LGMap is completely ignored, at least for the ghost communication.

The NaN values are from the later test, but I think it's good to focus on `test_ldc` first.

**Comment by dajuno** [comment]
Hm interesting. Curious that it works with 2 procs

**Comment by Sbte** [comment]
It's probably because it gets cut in half in such a way that all indices are continuous. That doesn't happen if you divide it into four quarters.

**Comment by dajuno** [comment]
Hm, not sure how to reproduce.

> What I tried now was to run the first ldc test on 4 cores. Then I printed the state vector and the indices (real and ghost) and compared. The state just has increasing numbers in it, so the indices should match the values (-1). They don't, which is what is causing issues. I'm quite sure that the reason is that the LGMap is completely ignored, at least for the ghost communication.

This seems to work for me:
```python
    for iloc, iglob in enumerate(interface.map.indices):
        with state.localForm() as lf:
            print(f"#{interface.comm.rank}", iglob, lf[iloc])
            assert iglob == lf[iloc] - 1
```

Actually, before this, I got an error with `A.getRow(i)` function in `extract_sorted_row`, which fails on procs that don't own `i`. Had to wrap the function body in    ` if i in range(*A.getOwnershipRange()): ... else: return [], []`.

**Comment by Sbte** [comment]
Right, but this is because the state vector that gets used to create the jacobian is wrong in the first place. You can print the state from the `from_array` method using `lf.array` after calling `ghostUpdate`, and you can see that there's duplicate values in there, which should be impossible.

**Comment by dajuno** [comment]
I see, you're right. I'll have a look

**Comment by dajuno** [comment]
I think the reason is this: https://petsc.org/main/manual/mat/#matrix-and-vector-layouts-and-storage-locations

> PETSc vectors always have a contiguous range of vector entries stored on each MPI rank. The first rank has entries from 0 to rend1 - 1, the next rank has entries from rend1 to rend2 - 1, etc.

**Comment by Sbte** [comment]
Well, then I guess we have to make our own "map" that maps between domain indices and petsc indices. Initially, you can just make an array that contains all global indices obtained by looping over the domains, and later this can be optimised so it only contains local and ghost nodes.

**Comment by dajuno** [comment]
PETSc has a tool for this. I'll check it out.

**Comment by dajuno** [comment]
A future problem to think about: when using PETSc with complex numbers, all PETSc vectors and matrices will be complex. I think this means that the numpy arrays created by `Discretization` must also be complex.

**Comment by Sbte** [comment]
I imagine we'll just have to handle the conversion in the PETSc interface.

**Comment by dajuno** [comment]
All implemented tests now succeed on multiple cores (tested with 1, 2, 4).
The tests include some trivial checks which helped me with the implementation, so I didn't delete them.

File: tests/test_PETSc.py [hunk 4]
```diff
@@ -68,12 +68,14 @@ def test_ldc():
     parameters = {"Reynolds Number": 100}
     n = nx * ny * nz * dof
 
+    # FIXME why not state = numpy.arange(1, n+1) ?
```
Comment (Sbte) [inline tests/test_PETSc.py]: Can you please remove the intermediate comments and disabled tests? Presumably they still pass in serial mode.

File: tests/test_PETSc.py [hunk 5]
```diff
@@ -136,18 +143,20 @@ def test_PETSc(nx=4):
...
     n = nx * ny * nz * dof
-    x0 = PETScInterface.Vector().createMPI(n)
+    x0 = PETScInterface.Vector.from_array(
+        interface.map, numpy.zeros(n, dtype=numpy.int32), ghosts=interface.ghosts
+    )
     x0 = continuation.newton(x0)
 
     start = 0
-    target = 2000
+    target = 100
```
Comment (Sbte) [inline tests/test_PETSc.py:341]: This changes the test, which is most likely undesired.

File: transiflow/interface/PETSc.py [hunk 6]
```diff
@@ -81,7 +129,7 @@ def solve(self, jac, rhs):
         ksp.setType("preonly")
         pc = ksp.getPC()
         pc.setType("lu")
-        pc.setFactorSolverType("mumps")
+        pc.setFactorSolverType("superlu_dist")
```
Comment (Sbte) [inline transiflow/interface/PETSc.py]: This probably desires a separate commit.

File: tests/test_PETSc.py [hunk 7]
```diff
@@ -34,16 +34,16 @@ def read_matrix(fname):
...
 
     values = []
     dirname = os.path.dirname(__file__)
     with open(os.path.join(dirname, fname), "r") as f:
-        for i, v in enumerate(f.readlines()):
+        for v in f.readlines():
             values.append(float(v.strip()))
 
-    vec = PETScInterface.Vector.from_array(values)
+    vec = PETScInterface.Vector.from_array(m, numpy.array(values, dtype=numpy.int32))
```
Comment (Sbte) [inline tests/test_PETSc.py]: Fixing the tests should go in the same commit as the implementational changes

File: tests/test_PETSc.py [hunk 8]
```diff
@@ -3,8 +3,7 @@
 import numpy
 import pytest
 
-from transiflow import Continuation, utils
-from transiflow.interface import SciPy as SciPyInterface
```
Comment (Sbte) [inline tests/test_PETSc.py]: This was added in the previous commit. The commits should probably be reversed.

File: tests/test_PETSc.py [hunk 9]
```diff
@@ -55,8 +55,8 @@ def extract_sorted_row(A, i):
 
 
 def extract_sorted_local_row(A, i):
-    indices = A.jcoA[A.begA[i] : A.begA[i + 1]]
-    values = A.coA[A.begA[i] : A.begA[i + 1]]
+    indices = A.jcoA[A.begA[i]: A.begA[i + 1]]
+    values = A.coA[A.begA[i]: A.begA[i + 1]]
```
Comment (Sbte) [inline tests/test_PETSc.py:66]: Please include flake8 fixes in the commit that broke it.

File: transiflow/interface/PETSc.py [hunk 10]
```diff
@@ -121,7 +121,7 @@ def solve(self, jac, rhs):
         ksp.setType("preonly")
         pc = ksp.getPC()
         pc.setType("lu")
-        pc.setFactorSolverType("superlu_dist")
+        pc.setFactorSolverType("mumps")
```
Comment (Sbte) [inline transiflow/interface/PETSc.py]: You changed this in an earlier commit.

File: transiflow/interface/PETSc.py [hunk 11]
```diff
@@ -38,61 +36,105 @@ class Interface(ParallelBaseInterface):
...
-        vec = Vector().createGhost(self.ghosts, self.size_global, comm=self.comm)
-        vec.setLGMap(self.map)
-        return vec
+        return Vector().createGhost(self.ghosts, (len(self.map.indices), None), comm=self.comm)
 
     def vector_from_array(self, array):
-        return Vector.from_array(self.map, array, self.ghosts)
+        return Vector.from_array(self.map, array[self.map_natural.indices], ghosts=self.ghosts)
+
+    def ghosts_petsc(self, ghosts_natural):
```
Comment (Sbte) [inline transiflow/interface/PETSc.py:80]: The changes in this commit should probably be split into several commits, but merged with previous commits that mention they "fix" the parallel implementation.

**Review by Sbte — CHANGES_REQUESTED** [review]
I added some comments. Mostly the commits should be such that every commit passes the tests, and every commit only contains a single functional change.

I just merged the MR that parameterized the continuation tests. Please add PETSc to the backend list for tests that work. This should be all tests that do not require an eigenvalue solver. Doing this probably requires the implementation of the `array_from_vector()` method which does an all gather. This will also be used in the near future for making all of the examples backend agnostic, meaning you can run the examples with any of the backends without additional changes.

File: tests/test_PETSc.py [hunk 12]
```diff
@@ -136,14 +327,14 @@ def test_PETSc(nx=4):
...
+    parameters = {"Reynolds Number": 0, "Verbose": True}
 
     interface = PETScInterface.Interface(parameters, nx, ny, nz, dim, dof)
 
     continuation = Continuation(interface, parameters)
 
-    n = nx * ny * nz * dof
-    x0 = PETScInterface.Vector().createMPI(n)
+    x0 = interface.vector()
+    x0.zeroEntries()
```
Comment (Sbte) [inline tests/test_PETSc.py:337]: I think my code assumes `interface.vector()` sets it to zero.

**Review by Sbte — CHANGES_REQUESTED** [review]
This is looking much better. Just some small comments.

File: tests/test_PETSc.py [hunk 13]
```diff
@@ -113,15 +296,23 @@ def test_norm():
...
-    state_dist = PETScInterface.Vector.from_array(state)
-    assert utils.norm(state) == utils.norm(state_dist)
+    interface = PETScInterface.Interface(parameters, nx, ny, nz, dim, dof)
+
+    state_dist = PETScInterface.Vector.from_array(
+        interface.map, state[interface.map_natural.indices]
+    )
+    state_dist2 = interface.vector_from_array(state)
+
+    assert utils.norm(state) == utils.norm(state_dist) == utils.norm(state_dist2)
```
[the remark this answers is not in the document]
Comment (dajuno) [reply tests/test_PETSc.py:315]: I don't know that this was introduced by a PEP. https://docs.python.org/3/reference/expressions.html#comparisons

[the remark this answers is not in the document]
Comment (Sbte) [reply tests/test_PETSc.py]: I see. You learn something new every day. There was no need to change it btw, I was just surprised :D.

**Review by Sbte — APPROVED** [review]

**Comment by Sbte** [comment]
Can you rebase on top of current master? I think that will fix the tests.

File: tests/test_PETSc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dajuno) [reply tests/test_PETSc.py]: hehe, comparison chaining is pretty convenient for cases like `if a < x < b: ...`

**Comment by dajuno** [comment]
> Can you rebase on top of current master? I think that will fix the tests.

done, but still not passing

**Comment by Sbte** [comment]
> done, but still not passing

Some were fixed, but not all. Seems like I have to push the same thing to JaDaPy.

**Comment by dajuno** [comment]
btw, it wouldn't be difficult to create a docker image for ubuntu with petsc installed, if we want to test petsc in the CI

**Comment by Sbte** [comment]
> btw, it wouldn't be difficult to create an ubuntu docker image with petsc installed, if we want to test petsc in the CI

Yeah, I was gonna do that next, but feel free to give it a go. I didn't do it for Trilinos because it's basically impossible.

**Comment by dajuno** [comment]
> > btw, it wouldn't be difficult to create an ubuntu docker image with petsc installed, if we want to test petsc in the CI
> 
> Yeah, I was gonna do that next, but feel free to give it a go. I didn't do it for Trilinos because it's basically impossible.

I made a first basic version here: https://github.com/dajuno/petsc-docker
All tests pass in the container, PETSc tests also in parallel

**Pull request merged**